src/components/ZoomRect/index.js

```diff
@@ -0,0 +1,341 @@
+  subDomainsByItemId: GriffPropTypes.subDomainsByItemId.isRequired,
+};
+
+const defaultProps = {
+  onMouseDown: null,
+  onMouseUp: null,
+  onMouseMove: null,
+  onMouseOut: null,
+  onClick: null,
+  onDoubleClick: null,
+};
+
+class ZoomRect extends React.Component {
+  componentDidMount() {
+    const { width, height } = this.props;
+    this.zoom = d3
+      .zoom()
+      .translateExtent([[0, 0], [width, height]])
+      .extent([[0, 0], [width, height]]);
+    this.rectSelection = d3.select(this.zoomNode);
+    this.syncZoomingState();
+  }
+
+  componentDidUpdate(prevProps) {
+    const { zoomAxes: prevZoomAxes } = prevProps;
+    const { zoomAxes: currZoomAxes } = this.props;
+    if (!isEqual(prevZoomAxes, currZoomAxes)) {
+      this.syncZoomingState();
+    }
+  }
+
+  onTouchStart = () => {
+    const {
+      event: { touches },
+    } = d3;
+    if (touches.length === 1) {
+      const [touch] = touches;
+      const { pageX: x, pageY: y } = touch;
+      this.lastTouch = {
+        time: x,
+        x,
+        y,
+      };
+    } else if (touches.length === 2) {
+      const [touchOne, touchTwo] = touches;
+      this.lastDeltas = {
+        time: Math.abs(touchOne.pageX - touchTwo.pageX),
+        x: Math.abs(touchOne.pageX - touchTwo.pageX),
+        y: Math.abs(touchOne.pageY - touchTwo.pageY),
+      };
+    }
+  };
+
+  // TODO: A lot of this is duplicated with `zoomed` -- maybe they can be
+  // consolidated?
+  onTouchMove = () => {
+    const { width, height } = this.props;
+
+    const distances = {
+      time: width,
+      x: width,
+      y: -height,
+    };
+
+    const {
+      event: { touches },
+    } = d3;
+    let updates = null;
+    if (touches.length === 1) {
+      // If there was only one touch, then it was a drag event.
+      updates = this.performTouchDrag(touches, distances);
+    } else if (touches.length === 2) {
+      // If there were two, then it is a zoom event.
+      updates = this.performTouchZoom(touches, distances);
+    } else {
+      // We don't support more complicated gestures.
+    }
+    if (updates) {
+      this.props.updateDomains(updates);
+    }
+  };
+
+  onTouchEnd = () => {
+    const {
+      event: { touches },
+    } = d3;
+    if (touches.length === 0) {
+      this.lastTouch = null;
+    } else if (touches.length === 1) {
+      this.lastDeltas = null;
+      const [touch] = touches;
+      const { pageX: x, pageY: y } = touch;
+      this.lastTouch = {
+        time: x,
+        x,
+        y,
+      };
+    }
+  };
+
+  performTouchDrag = (touches, distances) => {
+    const { itemIds, subDomainsByItemId, zoomAxes } = this.props;
+    const [touch] = touches;
+    const newTouchPosition = {
+      time: touch.pageX,
+      x: touch.pageX,
+      y: touch.pageY,
+    };
+    const updates = {};
+    itemIds.forEach(itemId => {
+      updates[itemId] = {};
+      ['x', 'y', 'time'].filter(axis => zoomAxes[axis]).forEach(axis => {
```

Comment:
I see `x`, `y` and `time` referenced around a lot - should these be moved to something like `AXIS_TYPES`?

Reply:
Yeah, they definitely should be -- will do!

Reply:
Done with the latest commit. Please take a look (you only need to view changes from that diff if you want to)

Reply:
SGTM 👍